### PR TITLE
Fix BashOperator on Windows

### LINKED ISSUE #41063: BashOperator Windows compatibility
### Apache Airflow version

2.9.3

### If "Other Airflow 2 version" selected, which one?

_No response_

### What happened?

When attempting to run the simple tutorial from:

https://airflow.apache.org/docs/apache-airflow/stable/tutorial/fundamentals.html

on a Windows machine,  there are two minor issues preventing this from working on Windows (related to  https://github.com/apache/airflow/issues/10388).

I resolved these issues locally and it seems to work on Windows now.

in airflow/hooks/subprocess.py line 77 in run_command() method:

```
            self.sub_process = Popen(
                command,
                stdout=PIPE,
                stderr=STDOUT,
                cwd=cwd,
                env=env if env or env == {} else os.environ,
                preexec_fn=pre_exec,
            )
```

fix `#1`: if running under Windows, set `preexec_fn=None`.

fix `#2`: for Popen call, we also need to pass in `close_fds=True`. On Windows, the code as is, will result in a race condition where the sub_process has not fully released the directory created in `cwd = stack.enter_context(TemporaryDirectory(prefix="airflowtmp"))`, and we'll get an exception:

```
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.12_3.12.1264.0_x64__qbz5n2kfra8p0\Lib\shutil.py", line 637, in _rmtree_unsafe
    os.rmdir(path)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\xxx\\AppData\\Local\\Temp\\airflowtmpngs54_vn'
```

### What you think should happen instead?

BashOperator should work on Windows. Airflow 2.10 addresses a few other blocking issues.

### How to reproduce

Running the simple tutorial at https://airflow.apache.org/docs/apache-airflow/stable/tutorial/fundamentals.html on a Windows box in Pycharm a few times will result in the exception.

### Operating System

Windows 11

### Versions of Apache Airflow Providers

_No response_

### Deployment

Virtualenv installation

### Deployment details

_No response_

### Anything else?

Race condition exception occurs 95% of the time when attempting to run the tutorial.

### Are you willing to submit PR?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by SledgeHammer01**
This PR is part of #10388. This fixes BashOperator for Windows as reported on #41063.
closes: #41063
- `preexec_fn` is not supported on Windows
- ~~close_fds=True on Windows, otherwise, we can get a race condition where the process is still holding onto the directory resulting in an exception:~~

```
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.12_3.12.1264.0_x64__qbz5n2kfra8p0\Lib\shutil.py", line 637, in _rmtree_unsafe
    os.rmdir(path)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\xxx\\AppData\\Local\\Temp\\airflowtmpngs54_vn'

```

- `TemporaryDirectory` will call `shutil.rmtree()` internally when the context exits. On Windows, `shutil.rmtree()` is unreliable and there is a race condition where even after `self.sub_process.wait()`, the process will still be holding onto the directory causing an exception. The work-around is to call `shutil.rmtree()` in a retry loop. If we're not running under Windows, `shutil.rmtree()` is reliable and no retry loop is needed.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by SledgeHammer01** [comment]
@potiuk can you take a look? For some reason, I'm unable to add reviewers.

**Comment by potiuk** [comment]
> @potiuk can you take a look? For some reason, I'm unable to add reviewers.

Yes. Reviewers will take a look when they have time and interest generally, not when they are called.

**Comment by potiuk** [comment]
There are some static checks failing.

**Comment by uranusjr** [comment]
Instead of having a giant try-except block, it would be much preferred to do a short feature detection before hand, and run one of the options depending on the detection result.

**Comment by SledgeHammer01** [comment]
> Instead of having a giant try-except block, it would be much preferred to do a short feature detection before hand, and run one of the options depending on the detection result.

I'm not quite sure what you mean. Can you explain further? Basically, the issue is known unreliability with `TemporaryDirectory` under Windows which occurs within the context of `contextLib::stack`. All the code for POSIX/Windows is 100% identical (except for `preexec_fn` not being supported on Windows.

`TemporaryDirectory` tear down doesn't occur until the context exits and the context tear down occurs and that's when the exception occurs on Windows. And all the logic for the method occurs inside of the context which is basically the enter/exit of `Popen`.

Also, PLEASE keep in mind, that the reason for so many commits on a "trivial" fix is because we're limited to Python 3.8. If you check the earlier commits, there ARE slightly nicer ways to do this, but unfortunately, those are not supported until Python 3.10+.

Are you suggesting to essentially write a `run_command_posix` and a `run_command_windows` that's a 100% copy & paste except for the retry loop? I did consider that, but personally ruled it out as a maintenance headache. People who want to run Airflow on Windows is a small community and thus Windows support is a low priority. Having two versions would mean people would have to update test / both versions.

File: airflow/hooks/subprocess.py [hunk 1]
```diff
@@ -61,41 +64,68 @@ def run_command(
...
+                    raise RuntimeError("The subprocess should be created here and is None!")
+                if self.sub_process.stdout is not None:
+                    for raw_line in iter(self.sub_process.stdout.readline, b""):
+                        line = raw_line.decode(output_encoding, errors="backslashreplace").rstrip()
+                        self.log.info("%s", line)
+
+                self.sub_process.wait()
+                self.log.info("Command exited with return code %s", self.sub_process.returncode)
+
+                return_code: int = self.sub_process.returncode
```
Comment (uranusjr) [inline airflow/hooks/subprocess.py]: What I mean is this block is much much too big. A permission error can happen anywhere, and you may catch what you don’t want accidentally. A try block should be more localised and contain the lines you expect the exception to be raised from.

If there are multiple lines that can raise an exception scattered in wide range of lines (but not all lines in the range raise), it is preferred to do a quick check before any of the code runs and branch out depending on the check result.

Comment (SledgeHammer01) [reply airflow/hooks/subprocess.py]: @uranusjr I updated to a simplified approach.

Comment (SledgeHammer01) [reply airflow/hooks/subprocess.py]: @uranusjr fixed mypy issue.

**Pull request closed without merging**